designs/0004-sparse-matrices.md

```diff
@@ -0,0 +1,390 @@
+
+Below we go over how sparse matrices can be constructed and operated on within each stan block.
+
+
+## Data
+
+Sparse matrix types in the Stan language can be constructed in the data block via [Coordinate List](https://en.wikipedia.org/wiki/Sparse_matrix#Coordinate_list_(COO)) notation using the row and column sizes, non-empty row/column indices, and values for those index positions. The non-zero (NZ) row and column indices are brought in as bounds in the `[]`.
+
+```stan
+data {
+int N; // Rows
+int M; // Cols
+int K; // number non-empty
+int nz_row_ind[K]; // Non-empty row positions
+int nz_col_ind[K]; // Non-empty col positions
+
+sparse_matrix[nz_row_ind, nz_col_ind, N, M] A
+}
+```
+
+With proper IO the data sparsity pattern can be brought in automatically via a json or rdump list of lists.
+
+```stan
+data {
+int N; // Rows
+int M; // Cols
+sparse_matrix[N, M] A
+}
+```
+
+Though not in the initial implementation, we can also add sparse vectors.
+
+```stan
+int N; // Rows
+int K; // number non-empty
+int nz_row_ind[K]; // Non-empty row positions
+// Can we do this?
+sparse_vector[nz_row_ind, N] B;
+```
+
+The above sparse matrix example with values
+
+```stan
+N = 5
+M = 5
+K = 7
+// Column major order
+nonzero_row_index[K] = [ 2, 3, 1, 3,  5, 3,  2, 5]
+nonzero_col_index[K] = [ 1, 1, 2, 2,  3, 4,  5, 5]
+val[K] =               [22, 7, 3, 5, 14, 1, 17, 8]
+```
+
+Would have the dense form of
+
+| col/row | 1  | 2 | 3  | 4 | 5  |
+|---------|----|---|----|---|----|
+| 1       | 0  | 3 | 0  | 0 | 0  |
+| 2       | 22 | 0 | 0  | 0 | 17 |
+| 3       | 7  | 5 | 0  | 1 | 0  |
+| 4       | 0  | 0 | 0  | 0 | 0  |
+| 5       | 0  | 0 | 14 | 0 | 8  |
+
+## Transformed Data
+
+Sparse matrices in this block can be defined dynamically and declared such as
+
+```stan
+transformed data {
+// Could construct here as well
+sparse_matrix[nz_row_ind, nz_col_ind, N, M] A =
+   to_sparse_matrix(nz_row_ind, nz_col_ind, N, M, vals);
+
+// Linear Algebra is cool
+sparse_matrix C = A * A';
+}
+```
+
+Because the sparsity pattern is given in the bounds `<>` the above multiplication result `C` will have it's own sparsity pattern deduced from the result of `A * A'`. This is a concern of Eigen and Stan-math and I'm pretty sure would not need anything particular in the stan compiler.
+
+## Parameters, Transformed Parameters, and Generated Quantities
+
+Parameters can be defined as above for data or deduced from the output of other functions.
+
+```stan
+parameters {
+  // Defining sparse matrices in parameters needs the non-zero elements
+  sparse_matrix[nz_row_ind, nz_col_ind, N, M] foo;
+}
+transformed parameters {
+  // Non-zero elements are deduced by the operation on x
+  sparse_matrix[N, N] K = cov_exp_quad(x, alpha, rho);
```

Comment:
cov_exp_quad is a bad example as it will always produce a dense matrix. There are some covariance functions which produce sparse covariance, but they are not useful in Stan as the sparsity structure depends on length scale parameter. Better example could be e.g.
```
sparse_matrix[N, M] P = A * alpha;
```